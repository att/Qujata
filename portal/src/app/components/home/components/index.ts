export * from './experiment';
export * from './latest-experiments';
