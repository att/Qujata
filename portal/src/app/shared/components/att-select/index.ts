export * from './AttSelect';
export * from './AttSelect.model';
