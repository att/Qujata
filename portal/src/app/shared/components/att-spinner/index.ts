export * from './Spinner';
export * from './Spinner.model';
