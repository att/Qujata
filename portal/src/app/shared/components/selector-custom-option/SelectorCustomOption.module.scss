@import "src/styles/variables-keys";

.algorithms_input_option {
  margin-inline-end: 10px;
  margin-inline-start: 25px;
  cursor: pointer;
}

.algorithms_input_option_title {
  margin-inline-end: 10px;
  cursor: pointer;
}

.option_wrapper {
  inline-size: 17px;
  block-size: 17px;
  margin-inline-end: 12px;
}

.iterations_input_option {
  margin-inline-end: 10px;
  cursor: pointer;
}

.input_option {
  display: none;
}