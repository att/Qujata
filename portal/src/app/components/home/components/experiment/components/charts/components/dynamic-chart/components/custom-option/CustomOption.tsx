import { GroupBase, OptionProps, components } from "react-select";
import styles from './CustomOption.module.scss';
import { AttSelectOption } from "../../../../../../../../../../shared/components/att-select";
import { capitalizeFirstLetter, getIconByValue } from "../../utils/dynamic-chart.utils";

export type SelectorCustomOptionProps = OptionProps<AttSelectOption<any>, true, GroupBase<AttSelectOption<any>>> & {
    onOptionChanged: (option: AttSelectOption) => void;
    showInputOption: boolean;
    setShowInputOption: (show: boolean) => void;
    inputValue: string;
    setInputValue: (value: string) => void;
    setMenuIsOpen: (isOpen: boolean) => void;
};

export const CustomOption: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
    const option: any = props.data;

    return (
        <components.Option {...props}>
            <img className={styles.icon} src={getIconByValue(option.value)} alt={option.value} />
            <span>{capitalizeFirstLetter(option.value)}</span>
        </components.Option>
    );
};
