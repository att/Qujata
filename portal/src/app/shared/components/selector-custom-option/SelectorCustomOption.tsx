import { OptionProps, components } from 'react-select';
import styles from './SelectorCustomOption.module.scss';
import { AttSelectOption } from '../att-select';
import { algorithmSections } from '../../../components/protocol-query/constants';
import cn from 'classnames';
import CheckedSvg from '../../../../assets/images/checked.svg';
import UnCheckedSvg from '../../../../assets/images/unchecked.svg';

export type SelectorCustomOptionProps = OptionProps<AttSelectOption, true> & {
  onOptionChanged: (option: AttSelectOption) => void;
};

export const AlgorithmsSelectorCustomOption: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
  const isSectionTitle = algorithmSections.includes((props.data as AttSelectOption).value);
  const optionStyle = isSectionTitle ? styles.algorithms_input_option_title : styles.algorithms_input_option;

  return (
    <components.Option {...props}>
      <div className={styles.option_wrapper}>
        <input
          type="checkbox"
          className={styles.input_option}
          checked={props.isSelected}
          onChange={() => props.onOptionChanged} />
          <img className={optionStyle} src={props.isSelected ? CheckedSvg : UnCheckedSvg} alt="checked" />
      </div>
      <span className={optionStyle}>{props.label}</span>
    </components.Option>
  );
};

export const IterationsSelectorCustomOption: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
  return (
    <components.Option {...props}>
      <div className={styles.option_wrapper}>
        <input
          type="checkbox"
          className={cn(styles.iterations_input_option, styles.input_option)}
          checked={props.isSelected}
          onChange={() => props.onOptionChanged} />
          <img src={props.isSelected ? CheckedSvg : UnCheckedSvg} alt="checked" />
      </div>
      <span>{props.label}</span>
    </components.Option>
  );
};