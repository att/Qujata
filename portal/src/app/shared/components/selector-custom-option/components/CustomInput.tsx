import { useCallback, useEffect, useMemo, useRef } from "react";
import styles from './CustomInput.module.scss';
import cn from 'classnames';
import { AttSelectOption } from "../../att-select";
import { SelectorCustomOptionProps } from "../SelectorCustomOption";
import { components } from "react-select";
import CheckedSvg from '../../../../../assets/images/checked.svg';
import UnCheckedSvg from '../../../../../assets/images/unchecked.svg';
import CleanSvg from '../../../../../assets/images/clean.svg';
import { Button, ButtonActionType, ButtonSize, ButtonStyleType } from "../../att-button";
import { SELECTOR_CUSTOM_OPTION_EN } from "../translate/en";

const CheckedAriaLabel: string = 'checked';
const CleanAriaLabel: string = 'clean';

type OnEventInputChange = (event: React.ChangeEvent<HTMLInputElement>) => void;
type OnEventHandler = () => void;

export const CustomInput: React.FC<SelectorCustomOptionProps> = (props: SelectorCustomOptionProps) => {
  const { showInputOption, setShowInputOption, inputValue, setInputValue, setMenuIsOpen } = props;
  const inputOption = useMemo<AttSelectOption>(() => ({ label: inputValue, value: inputValue, metadata: { isInput: true } }), [inputValue]);
  const inputRef = useRef<HTMLInputElement>(null);
  const isInputOptionExists = (props.options as AttSelectOption[]).some((option: AttSelectOption) => option.metadata?.isInput);

  const handleInputChange: OnEventInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value;
    setInputValue(newValue);
    if (isNaN(Number(newValue))) {
      props.selectOption(inputOption as AttSelectOption);
    }
  };

  const applyCustomOption: OnEventHandler = useCallback((): void => {
    delete props.data.metadata;
    setShowInputOption(false);
    setTimeout(() => props.selectOption({ label: inputOption.label, value: inputOption.value }), 0);
  }, [props, setShowInputOption, inputOption]);

  useEffect(() => {
    if (showInputOption && props.data.metadata?.isInput) {
      inputRef.current?.focus();
      props.data.label = props.data.value = inputValue;
    }
  }, [showInputOption, props.data, inputValue]);

  useEffect(() => {
    const currentOptionsSelected = (props.getValue() ?? []).map((option: AttSelectOption) => option.label);
    const lastOptionSelected = currentOptionsSelected[currentOptionsSelected.length - 1];

    // handles deselection when the user selects the input option and then selects another option
    if (currentOptionsSelected.includes(inputOption.label) && lastOptionSelected !== inputOption.label && props.data.metadata?.isInput) {
      props.selectOption(inputOption as AttSelectOption);
    }
    // handles deselection when the user selects the input option and then types a wrong value
    if (lastOptionSelected === inputOption.label && isNaN(Number(inputValue))) {
      props.selectOption(inputOption as AttSelectOption);
    }
  }, [props, inputOption, inputValue]);

  return (
    <>
      {showInputOption && props.data.metadata?.isInput && (
        <components.Option {...props}>
          <div className={styles.add_new_wrapper}>
            <div className={styles.add_new_checkbox_wrapper}>
              <input
                data-testid="add_new_input"
                id={props.label}
                type="checkbox"
                className={cn(styles.iterations_input_option, styles.input_option)}
                checked={props.isSelected}
                onChange={() => props.onOptionChanged}
              />
              <img src={props.isSelected ? CheckedSvg : UnCheckedSvg} alt={CheckedAriaLabel} />
            </div>
            <div className={styles.input_wrapper}>
              <input
                data-testid="custom_input"
                id={props.label}
                ref={inputRef}
                value={inputValue}
                onChange={handleInputChange}
                onFocus={() => setMenuIsOpen(true)}
                className={styles.add_new_input_option}
                disabled={!props.isSelected}
              />
              {inputValue !== '' && (
                <>
                  <img
                    data-testid="clean_icon"
                    className={styles.clean_icon}
                    onClick={() => setInputValue('')}
                    src={CleanSvg}
                    alt={CleanAriaLabel}
                    tabIndex={props.isSelected ? 0 : -1}
                  />
                  <Button
                    className={styles.add_button}
                    ariaLabel={props.label}
                    size={ButtonSize.NONE}
                    styleType={ButtonStyleType.WRAPPER}
                    actionType={ButtonActionType.SUBMIT}
                    onButtonClick={applyCustomOption}
                    disabled={!props.isSelected || isNaN(Number(inputValue))}
                  >{SELECTOR_CUSTOM_OPTION_EN.ADD_BUTTON}</Button>
                </>
              )}

            </div>
          </div>
        </components.Option>
      )}
      {props.data.metadata?.isAddNewButton && isInputOptionExists && (
        <Button
          className={styles.add_new_button}
          ariaLabel={props.label}
          size={ButtonSize.NONE}
          styleType={ButtonStyleType.WRAPPER}
          actionType={ButtonActionType.BUTTON}
          onButtonClick={() => setShowInputOption(true)}
          disabled={showInputOption}
        >{props.label}</Button>
      )}
    </>
  );
}