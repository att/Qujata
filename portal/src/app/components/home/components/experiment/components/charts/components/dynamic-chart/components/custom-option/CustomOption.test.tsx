import { render, RenderResult } from '@testing-library/react';
import { CustomOption } from './CustomOption';
import { SelectorCustomOptionProps } from '../../../../../../../../../../shared/components/selector-custom-option';

describe('CustomOption', () => {
  const mockOption = { value: 'option1', label: 'Option 1' };
  const mockProps: SelectorCustomOptionProps = {
    data: mockOption,
    isSelected: false,
    selectOption: jest.fn(),
    label: 'Option 1',
    innerProps: {},
    innerRef: jest.fn(),
    children: null,
    type: 'option',
    isDisabled: false,
    isFocused: false,
    clearValue: jest.fn(),
    cx: jest.fn(),
    getStyles: jest.fn(),
    getClassNames: jest.fn(),
    getValue: jest.fn().mockReturnValue([{ label: 'Option 1', value: 'option1' }]),
    hasValue: true,
    isMulti: true,
    isRtl: false,
    options: [],
    selectProps: expect.any(Object),
    setValue: jest.fn(),
    theme: expect.any(Object),
    onOptionChanged: jest.fn(),
    showInputOption: false,
    setShowInputOption: jest.fn(),
    inputValue: '1111',
    setInputValue: jest.fn(),
    setMenuIsOpen: jest.fn(),
  };

  it('should render CustomOption', () => {
    const { container }: RenderResult = render(<CustomOption {...mockProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
