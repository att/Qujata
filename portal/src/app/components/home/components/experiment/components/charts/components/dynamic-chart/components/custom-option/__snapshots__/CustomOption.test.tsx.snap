// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomOption should render CustomOption 1`] = `
<div
  aria-disabled="false"
  class="css-0"
>
  <img
    alt="option1"
    class="icon"
    src="bar.svg"
  />
  <span>
    Option1
  </span>
</div>
`;
